### ISSUE TITLE
Jellyfin Doesn't Transcode HEVC 4K to 1080p on Roku 4

### ISSUE THREAD

**Issue by NeuroDawg** [jellyfin-archive/jellyfin-roku-legacy#944]
**Software Versions**
Jellyfin Server Version: 10.8.8
Roku Client Version: 1.6 build 2
**Describe the bug**
On a Roku 4, which is 4K capable, but attached to a 1080p TV, Jellyfin transcoded HEVC 4K files to h264 4K, rather than transcoding to 1080p. Transcoding an HEVC 1080p file to h264 1080p works well. It appears that Jellyfin client isn't reporting the correct TV resolution to the server, resulting in the transcoded resolution staying unchanged. This may be related to [issue #722](https://github.com/jellyfin/jellyfin-roku/issues/722) which was marked as closed, but doesn't seem to address the issue. I don't think it's a bitrate issue, as I can play h264 1080p files that are 38Mbps (direct BD remuxes via makemkv) without any problem, so I don't trust the Roku chart mentioned in the discussion of #722.

**Expected behavior**
HEVC 4K content should be transcoded to h264 1080p content on a Roku 4 attached to a 1080p TV.

**Logs**
**Pertinent section from my transcode log:**
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> h264 (h264_nvenc))
  Stream #0:2 -> #0:1 (copy)
Press [q] to stop, [?] for help
Output #0, hls, to '/config/transcodes/27ffc89e272dd256a277151732cc2590.m3u8':
  Metadata:
    encoder         : Lavf59.27.100
  Stream #0:0: Video: h264 (High), cuda(tv, bt709, progressive), 3840x2160 [SAR 1:1 DAR 16:9], q=2-31, 98547 kb/s, 23.98 fps, 90k tbn
    Metadata:
      encoder         : Lavc59.37.100 h264_nvenc
    Side data:
      cpb: bitrate max/min/avg: 98547528/0/98547528 buffer size: 197095056 vbv_delay: N/A
  Stream #0:1: Audio: ac3, 48000 Hz, 5.1(side), fltp, 640 kb/s (default)

**Connection Information**
Server is local (LAN) and connection is http.

**Comment by jimdogx** [comment]
Issue #722 was just recently merged / closed and isn't in the latest build.  So the next release may address your issue.

**Comment by NeuroDawg** [comment]
> Issue #722 was just recently merged / closed and isn't in the latest build. So the next release may address your issue.

As I understand #722 and #909, which I may understand incorrectly, when this new code is released, my Jellyfin Roku client will report back to my Jellyfin server that there are bitrate limits that should be applied to the file that is streamed. This limits will be based upon a chart published by Roku. But I'm not sure how that's going to fix my need to have the client report to the server the display resolution of my TV to ensure that any 4K content is transcoded to 1080p content. I have my doubts that it will fix it at all. Here's why:

1. The Roku guidelines state that h264 files should have a max bitrate of "up to 10mbps" and that HEVC files should have a max bitrate of "up to 40mbps'.
2. These guidelines are not reality, as I know from experience my Roku 4 has no problems streaming an h264 movie that has a video bitrate of 38mbps.
3. If I have an HEVC 4K file that has a bitrate of 60mbps, this new code will see that it's HEVC and tell the server to limit the bitrate to <40mbps. However, in the discussion of 722 and 909 I see nothing that says my client will report the TV resolution of 1080p; so I anticipate that it will simply transcode HEVC 4K -> H264 4K to meet these false bitrate limits. This is what my system is doing now.

I will watch for the new release and report back.

**Comment by jimdogx** [comment]
> I will watch for the new release and report back.

👍

**Comment by jimdogx** [comment]
@NeuroDawg as an FYI, I recently acquired a movie that is MKV, H264 (High) 16 mbps + DTS audio.  The video transcodes as you'd expect.  So I did the following:

1. Removed the 10 mbps limit from my local version of the Roku client.
2. The video "direct played", but it was still being transcoded to 10 mbps because the audio codec wasn't supported.
3. I then used ffmpeg to switch the audio to mp3.
4. The server reported that the video *and* audio were now direct playing, however the Roku just displayed a blank screen.

Conclusion, unless the server transcodes the video down to 10 mbps (for whatever reason, be it video or audio), my Roku can't play it.

**Comment by NeuroDawg** [comment]
@jimdogx - That's interesting. Here are a couple of questions and a comment:

1. What was the video resolution of the H264 file you tested, 1080p or 4K (and can I assume your testing with 1080p TV)? If the video bitrate is 1080p, I wouldn't expect it to transcode the video, only the audio. If it's 4K and you're viewing on a 1080p TV, then I would expect transcoding.
2. If Jellyfin is reporting the video as "direct played" there shouldn't be any transcoding whatsoever. How do you say your file "direct played" but then say it was transcoding? My understanding is that "direct play" and "transcode" are mutually exclusive.
3. I get a blank screen as well (and can hear some transcoded audio) when I attempt to play an HEVC 4K mkv file on my Roku4. My log, as pasted above, clearly show that it's being transcoded HEVC -> H264, but not being transcoded 4K -> 1080p; it's staying in 4K. The video bitrate is reported as 98mbps. I don't know if I can't see the video because the bitrate is too high, or if it's because the video is in a resolution my TV can't handle (I don't think the Roku will transcode video down to 1080p).
4. How are you determining the bitrate that is being transmitted to the client? I've been looking at the "stream:" and "cpb:" lines in my transcode logs. Am I looking in the wrong place?
5.  I have directly played mkv files in which the video is 1080p HEVC and 1080p H264 and have bitrates higher than Roku says are the maximum. I didn't keep those logs when I was doing this testing last week. I'll play those files again and post the appropriate sections from my transcode log files.
6. I don't see a way in the Jellyfin client on my roku to limit streaming speeds. Am I just not seeing it, or is this something hard-coded into the client? If I can figure out a way to limit the streaming speed, I'd be happy to run tests and see what happens.

Thanks for your comments. I hope that all is solved by forcing bitrate restrictions on the clients.

**Comment by jimdogx** [comment]
1. 1080p video on a 1080p TV.  Transcoded because the bitrate was too high.

2. When I removed the bitrate limiter on my Roku client, it still transcodes because of the audio.  In theory, only the audio gets transcoded (you can see "ffmpeg -codec: v:0 copy" in the logs), however the video is being transcoded to HLS for streaming and looking at the individual transcoded files, I see variable bitrates (8 mbps, 10 mbps, up to 20 mbps and back down).  The Roku docs do say it can handle up to "1.5x average" for peak bitrates.  The Roku client is reporting 10 mbps.

TLDR: Even though "only" the audio is forcing the transcode, I believe that when the video is split in to individual .ts files, the variable bitrate is still ~10 mbps.

3. Press up while watching the video.  Pretty sure (from what I'm seeing) that it's because of the bitrate.  Not the resolution.

4.  Press the up arrow while viewing the video.

Note: There are two (possible) sections, "Transcoding info" and "Streaming info".  The Transcoding section only shows up if it's being transcoded.  And unfortunately there is a [bug](https://github.com/jellyfin/jellyfin-roku/issues/858) where the Streaming info isn't 100% reliable.

5.  I would double check by pressing up on the video.  I bet they are being lowered to 10 mbps.

6. It's hardcoded in the latest unstable branch.  I believe there was (or is) a request to make it a user setting.

Keep in mind that I am not an ffmpeg expert by any means, so all of this is just from me playing around and looking at the ffmpeg docs etc.

**Comment by jimdogx** [comment]
Here is the transcode info for the 16 mbps video that in theory is only transcoding the audio.  Notice the video is "direct", but it says 10 mbps.  Also notice that the Streaming section is bogus (that is the aforementioned bug).

![Screenshot 2023-01-10 at 6 20 04 AM](https://user-images.githubusercontent.com/18198517/211562489-4d3dd793-ee22-4dd2-8a38-911a7a93016e.png)

**Comment by jimdogx** [comment]
@NeuroDawg 1.6.3 is out and would be good to know if the bitrate limits in the new build address your issue.

**Comment by NeuroDawg** [comment]
Well, it may take me a bit to get back to you. It appears that my upgrade to 10.8.9 has broken my hardware acceleration. As soon as I get that sorted, I'll let you know.

**Comment by NeuroDawg** [comment]
Ok. The HW acceleration was fixed by simply re-creating my docker container. Don't know why HW acceleration failed after an upgrade, but it's working now.

With the new Roku client my 4K files are playing, and being transcoded down to 10.6Mbps h264 files. I'm assuming that this bitrate is so low, because that's what the Roku documents state can by handled by h264.

When I play an 1080p H264 movie, it also gets transcoded down to 10.6Mbps. This sucks, as previously my Roku was direct playing h264 files at 38-40Mbps without transcoding. So now I'm taking a quality hit for files that previously played just fine.

I think it's time to ditch my Roku for an Nvidia Shield (which is my main streamer in my living room; the Roku is in the guest bedroom). I doubt you all will return the Roku client to previous h264 bitrate limits (which were none, I believe). Maybe put that in the settings as a choice?

**Comment by jimdogx** [comment]
https://github.com/jellyfin/jellyfin-roku/pull/975

**Comment by WavesOfSalt** [comment]
I don't know if it ever worked, but since issue #722, the only change I've seen was from only playing audio to fully failing.
![image](https://user-images.githubusercontent.com/65592397/236640142-0595bc1e-9a88-40ab-bf64-2ef1f2054b0e.png)
